test/fixtures/dialects/bigquery/select_hyphenated_table_name_in_from.yml

```diff
@@ -19,29 +19,19 @@ file:
           from_expression_element:
             table_expression:
               hyphenated_object_reference:
```

Comment:
Interesting changes below. IIRC, `hyphenated_object_reference` was added to smooth over the way the lexer was splitting the names into pieces. I wonder if it's no longer needed.

Reply:
I'm not sure - possibly one for @tunetheweb

Reply:
I didn't implement this, but never been particularly happy the way it was implemented and been meaning to have a crack at this. So glad to see some clean ups here. Though wasn't exactly the way I was intending it so will do some tests that it doesn't cause problems. Will comment separately if I think it might...

Reply:
OK the problem with this change, is that `newcol1` in the below no longer parses (though `newcol2` does parse):

```sql
SELECT
   col1-col2 AS newcol1,
   col1 - col2 AS newcol2
FROM
  table-123
```

This is because you've added a lexer that swallows up `col1-col2` as a single identifier, when where it's actually intended to mean `col1 MINUS col2`.

Basically hyphenated identifiers are really valid in table names (and only in some parts of table names at that), so everywhere else this will be presumed to mean minus.

I presume is why the current "hacky" way of using a delimited segment was implemented, even if technically it incorrectly breaks up the identifier more than it should.

I've played around it some more  but can't find a cleaner way that what was implemented already, so think we need to revert to this.

On that, we should add above example SQL to `test/fixtures/dialects/bigquery/select_hyphenated_table_name_in_from.sql` so we catch anyone else attempting to simplify this like this in future to warn them why it's not possible.

Reply:
So how do we distinguish between a hyphenated identifier and a minus operation with no spaces? It's impossible to infer intent, no? Or are you saying the parser literally fails?

Reply:
oh, i see, there's a mistake in my lexer, it's cause there's a number before the `-`, this should fix it 
```
r"[a-zA-Z_][a-zA-Z_0-9]*[-][a-zA-Z_0-9-]*"
```

Reply:
The lexer can't. Hence why this can't be added to lexer (like you have here).

So yes in this branch the parser fails:

```
% echo "SELECT a-b FROM a-b;" | sqlfluff parse --dialect bigquery -
[L:  1, P:  1]      |file:
[L:  1, P:  1]      |    statement:
[L:  1, P:  1]      |        select_statement:
[L:  1, P:  1]      |            select_clause:
[L:  1, P:  1]      |                unparsable:                                   !! Expected: 'SelectClauseSegment'
[L:  1, P:  1]      |                    keyword:                                  'SELECT'
[L:  1, P:  7]      |                    whitespace:                               ' '
[L:  1, P:  8]      |                    raw:                                      'a-b'
[L:  1, P: 11]      |            whitespace:                                       ' '
[L:  1, P: 12]      |            [META] dedent:
[L:  1, P: 12]      |            from_clause:
[L:  1, P: 12]      |                keyword:                                      'FROM'
[L:  1, P: 16]      |                whitespace:                                   ' '
[L:  1, P: 17]      |                from_expression:
[L:  1, P: 17]      |                    [META] indent:
[L:  1, P: 17]      |                    from_expression_element:
[L:  1, P: 17]      |                        table_expression:
[L:  1, P: 17]      |                            hyphenated_object_reference:
[L:  1, P: 17]      |                                identifier:                   'a-b'
[L:  1, P: 20]      |                    [META] dedent:
[L:  1, P: 20]      |    statement_terminator:                                     ';'
[L:  1, P: 21]      |    newline:                                                  '\n'

==== parsing violations ====
L:   1 | P:   1 |  PRS | Line 1, Position 1: Found unparsable section: 'SELECT a-b'
```

Whereas in main it doesn't:

```
% echo "SELECT a-b FROM a-b;" | sqlfluff parse --dialect bigquery -
[L:  1, P:  1]      |file:
[L:  1, P:  1]      |    statement:
[L:  1, P:  1]      |        select_statement:
[L:  1, P:  1]      |            select_clause:
[L:  1, P:  1]      |                keyword:                                      'SELECT'
[L:  1, P:  7]      |                [META] indent:
[L:  1, P:  7]      |                whitespace:                                   ' '
[L:  1, P:  8]      |                select_clause_element:
[L:  1, P:  8]      |                    expression:
[L:  1, P:  8]      |                        column_reference:
[L:  1, P:  8]      |                            identifier:                       'a'
[L:  1, P:  9]      |                        binary_operator:                      '-'
[L:  1, P: 10]      |                        column_reference:
[L:  1, P: 10]      |                            identifier:                       'b'
[L:  1, P: 11]      |            whitespace:                                       ' '
[L:  1, P: 12]      |            [META] dedent:
[L:  1, P: 12]      |            from_clause:
[L:  1, P: 12]      |                keyword:                                      'FROM'
[L:  1, P: 16]      |                whitespace:                                   ' '
[L:  1, P: 17]      |                from_expression:
[L:  1, P: 17]      |                    [META] indent:
[L:  1, P: 17]      |                    from_expression_element:
[L:  1, P: 17]      |                        table_expression:
[L:  1, P: 17]      |                            hyphenated_object_reference:
[L:  1, P: 17]      |                                identifier:                   'a'
[L:  1, P: 18]      |                                dash:                         '-'
[L:  1, P: 19]      |                                identifier:                   'b'
[L:  1, P: 20]      |                    [META] dedent:
[L:  1, P: 20]      |    statement_terminator:                                     ';'
[L:  1, P: 21]      |    newline:                                                  '\n'
```

The parse can be more flexible though as it has context. So in `main` `HyphenatedObjectReferenceSegment` is only in `TableExpressionSegment` (where a minus wouldn't make sense) but not in other places an Expression (which could include a minus) are.

But, as I say, this only works because `HyphenatedObjectReferenceSegment` is made up of grammar chunks, and not lexer chunks.

Reply:
> oh, i see, there's a mistake in my lexer, it's cause there's a number before the `-`, this should fix it
> 
> ```
> r"[a-zA-Z_][a-zA-Z_0-9]*[-][a-zA-Z_0-9-]*"
> ```

No, this still won't work because we don't WANT it to lex `col1-col2` as a single chunk.

Reply:
Oh gotcha, will try and fix tonight 👍

Reply:
I think I added the original "hyphenated object" stuff long ago, and yes, this was the exact reason -- it's an ugly workaround for an otherwise unsolvable problem.

Reply:
Sorted